coreneuron/utils/reports/nrnreport.cpp

```diff
@@ -29,222 +29,371 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #include <iostream>
 #include <vector>
 #include <algorithm>
+#include <map>
+#include <set>
+#include "coreneuron/nrniv/nrn_assert.h"
+#include "coreneuron/nrniv/netcvode.h"
 #include "coreneuron/nrnoc/multicore.h"
 #include "coreneuron/utils/reports/nrnreport.h"
 #include "coreneuron/utils/reports/nrnsection_mapping.h"
-#include "coreneuron/nrniv/nrn_assert.h"
-
+#include "coreneuron/nrnoc/mech_mapping.hpp"
+#include "coreneuron/nrnoc/membfunc.h"
 #ifdef ENABLE_REPORTING
 #include "reportinglib/Records.h"
 #endif
 
-int step = 0;
-
-/** constructor */
-ReportEvent::ReportEvent(double t) {
-    dt = t;
-    step = 0;
-}
+/*
+ * ---                      ---
+ *     Forward Declarations
+ * ---                      ---
+ */
+#ifdef ENABLE_REPORTING
+class ReportEvent;
+#endif //ENABLE_REPORTING
 
-/** on deliver, call ReportingLib and setup next event */
 
-void ReportEvent::deliver(double t, NetCvode* nc, NrnThread* nt) {
-// avoid pgc++-Fatal-/opt/pgi/linux86-64/16.5/bin/pggpp2ex TERMINATED by signal 11
+/*
+ * ---                         ---
+ *     Internal Implementation
+ * ---                         ---
+ */
 #ifdef ENABLE_REPORTING
 
-/** @todo: reportinglib is not thread safe, fix this */
-// clang-format off
-    #pragma omp critical
-    // clang-format on
-    {
-// each thread needs to know its own step
-#ifdef ENABLE_REPORTING
-        // TODO: do not rebuild this cellid vector each reporting dt!
-        std::vector<int> temp;
-        for (int i = 0; i < nt->ncell; i++) {
-            PreSyn& presyn = nt->presyns[i];
-            temp.push_back(presyn.gid_);
+static  std::vector<ReportEvent*> reports;
+struct  VarWithMapping {
+  int     id;
+  double* var_value;
+  VarWithMapping (int id_, double* v_): id(id_), var_value(v_) {}
+};
+
+typedef std::map <int, std::vector<VarWithMapping> > VarsToReport;
+typedef std::map <int, std::vector<std::string> >    ReportFilter;
+
+class ReportEvent : public DiscreteEvent {
+  private:
+    double dt;
+    unsigned long step;
+    char report_name [MAX_REPORT_NAME_LEN];
```

Comment:
std::string? if reportinglib not taking const char* then change it

Reply:
I made interface in reporting lib as const char* (C interface) as I didnt assume it to be a c++ interface: everything in header is C interface.